doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1417,77 +1412,57 @@ Managing the local context
       :undocumented:
 
-   .. tacv:: set (@ident := @term) in @goal_occurrences
-
-      This notation allows specifying which occurrences of :token:`term` have
-      to be substituted in the context. The :n:`in @goal_occurrences` clause
-      is an occurrence clause whose syntax and behavior are described in
-      :ref:`goal occurrences <occurrenceclauses>`.
-
-   .. tacv:: set (@ident {* @binder } := @term) {? in @goal_occurrences }
-
-      This is equivalent to :n:`set (@ident := fun {* @binder } => @term) {? in @goal_occurrences }`.
-
-   .. tacv:: set @term {? in @goal_occurrences }
+   .. tacn:: eset @bindings_with_parameters {? @occurrences }
+             eset @one_term {? @as_name } {? @occurrences }
+      :name: eset; _
 
-      This behaves as :n:`set (@ident := @term) {? in @goal_occurrences }`
-      but :token:`ident` is generated by Coq.
+      .. insertprodn as_name as_name
 
-   .. tacv:: eset (@ident {* @binder } := @term) {? in @goal_occurrences }
-             eset @term {? in @goal_occurrences }
-      :name: eset; _
+      .. prodn::
+         as_name ::= as @ident
 
-      While the different variants of :tacn:`set` expect that no existential
-      variables are generated by the tactic, :tacn:`eset` removes this
-      constraint. In practice, this is relevant only when :tacn:`eset` is
+      Similar to :tacn:`set`, but generates an existential variable when
+      a variable can't be instantiated instead of failing.
+      In practice, this is relevant only when :tacn:`eset` is
       used as a synonym of :tacn:`epose`, i.e. when the :token:`term` does
       not occur in the goal.
 
-.. tacn:: remember @term as @ident__1 {? eqn:@naming_intropattern }
-   :name: remember
+.. tacn:: remember @one_term {? @as_name } {? eqn : @naming_intropattern } {? in @goal_occurrences }
 
-   This behaves as :n:`set (@ident := @term) in *`, using a logical
-   (Leibniz’s) equality instead of a local definition.
+   Equivalent to :n:`set (@ident := @term) in *` but uses :term:`Leibniz equality`
+   instead of a :term:`local definition <context-local definition>`.
    Use :n:`@naming_intropattern` to name or split up the new equation.
 
-   .. tacv:: remember @term as @ident__1 {? eqn:@naming_intropattern } in @goal_occurrences
-
-      This is a more general form of :tacn:`remember` that remembers the
-      occurrences of :token:`term` specified by an occurrence set.
-
-   .. tacv:: eremember @term as @ident__1 {? eqn:@naming_intropattern } {? in @goal_occurrences }
-      :name: eremember
+   .. tacn:: eremember @one_term {? @as_name } {? eqn : @naming_intropattern } {? in @goal_occurrences }
 
-      While the different variants of :tacn:`remember` expect that no
-      existential variables are generated by the tactic, :tacn:`eremember`
-      removes this constraint.
+      Similar to :tacn:`remember`, but generates an existential variable when
+      a variable can't be intantiated rather than causing the tactic to fail.
```

Comment:
Similar comment to above.

Reply:
Also in `epose`.  I probably should mention the `e*` tactics as a group (`eapply` et al) in the existential variables section.

Reply:
I need to better understand when `e*` tactics introduce existential variables.  Look at the `eapply` in [this example](https://coq.gitlab.io/-/coq/-/jobs/3115218600/artifacts/_build/default/doc/refman-html/language/extensions/evars.html#automatic-resolution-of-existential-variables).  For `le_lt_trans`, it looks like `n < p` is unified with the goal and m is a new unbound variable, not unified, so it becomes an existential variable.  But I don't quite see why an EV must be introduced for `lt_irrefl` that requires introducing an existential variable.  Is it always related to some variable that's still unbound after unification (for all `e*` tactics)?

`Lemma le_lt_trans n m p : n <= m -> m < p -> n < p.`
`Theorem lt_irrefl : forall n, ~ (n < n).`

Reply:
Recall that `~ (n < n)` is defined as `n < n -> False`. In this proof, `lt_irrefl` is applied to a goal that is `False`. The conclusion of the lemma is unified to the goal, but that doesn't tell us which value `n` should be instantiated with. That's why the existential variable is introduced.

Reply:
I added a short section on `e*` tactics, which states "A number of tactics have companion tactics that create existential variables when a new, unspecified variable is added to the context rather than failing."

However, I think "when a new, unspecified variable is added to the context" is fairly vague.  It isn't very clear for `eintros`, and I have trouble imagining how the other `e*` tactics would create existential variables.  Perhaps you could provide a few terse examples that I could include or generalize from?  If there are precise, understandable rules I'd like to state them.

PS It took me 15 minutes to figure out how to invoke `Nat.lt_trans`, which is actually defined in `NZorder.v` and is available when you import `Arith`.  `Require Import Nat.` doesn't do the trick.  How do users know to import `Arith`??  `Search " _ > _"` is also underwhelming--off the top of my head, maybe you should be able to highlight part of the context and get a list of theorems that could be applied to it (taking into account whether it's the goal or a hypothesis that's highlighted).

Reply:
Note that I don't suggest using my example above (too convoluted). My impression is that what you wrote is already better than nothing. We should just highlight two tactics of the `e` set that are special:
- `eassumption` which resolves existential variables (which `assumption` refuses to do)
- `eauto` which internally uses `simple eapply` and `eassumption` but will not (usually) leave any existential variables (unless you do some advanced stuff with external hints).

We could also note (but maybe not worth) that since `constructor`, `exists`, `left`, `right`, `split` and `transitivity` are all just variants of `apply`, the same holds true for their `e` variants which are variants of `eapply`.

Reply:
Now "A number of tactics have companion tactics that create existential variables
when the base tactic would fail because of unbound variables."

I added a few words to `eassumption` and `eauto`.  Not motivated to fix the wording for every `e*` tactic in this PR.

> I don't know how people know. People just learn early on to do Require Import Arith and nothing else.
>
> Yes, Search could use lots of improvements. Plenty of work to do in this direction.

BTW, after we're done with this PR, I'd like to chat about starting a list of the most important improvements we'd like to see in Coq.

Reply:
>Now "A number of tactics have companion tactics that create existential variables
when the base tactic would fail because of unbound variables."

Maybe use "uninstantiated" instead of "unbound"?

Reply:
In the standard library, only the following are used: eapply eassert eassumption eauto econstructor edestruct enough (3 places) eexact (8 places) eexists epose erewrite esplit etransitivity

Not used: ecase ediscriminate eelim einduction einjection eintros eleft eremember eright eset esimplify_eq